### PR TITLE
User Form

### PR DESCRIPTION
**Pull request by ashleygranitto**
This pull request features functionality related to the user form. The primary features included are as follows:

1) user form: the form is currently basically styled. It was built using MWC and includes an automatic validation check for unreasonable ages (numbers entered not in the range 0..120) and validation checks for each input field for completeness upon pressing the "Save" button. 

2) cookie service: upon saving the user form, the cookie service is utilized to save each input as a cookie. The service also includes functionality to return the value given a cookie key, returning an empty string in the event that such a cookie does not exist. The latter functionality is utilized by the form as a means to auto-populate data upon loading the form.

3) "state management": various components were given custom event listeners that--when activated--update the state of the main app component. A function was added to render the appropriate components based on the application's current state. Now, when the user icon button is clicked on the recording page, the user is taken to the form, and when the user exits the form they are taken back to the record page. 

4) post request: the coookie service is utilized by the UtteranceApiService to include various pieces of user information in the  Utterance POST request (another pr was made for the backend component of this).

5) MWC: I installed several new material web components.

image

### CODE REVIEW

File: ui/src/VoxettaUserForm.js [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+                justify-content: center; 
+                text-align: center;  
+                width: 100%; 
+            }
+            h1.title {
+                font-family: 'Roboto';
+                font-size: 30px;
+            }
+            h2.description {
+                font-family: 'Roboto';
```
Comment (eldrickb) [inline ui/src/VoxettaUserForm.js:46]: might want to import Google Font in header index.html?

File: ui/src/VoxettaUserForm.js [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        `;
+    }
+
+    constructor() {
+        super();
+        this.cookieService = new CookieService();
+        this.userId = this.cookieService.getCookieValue("userId");
+        this.gender = this.cookieService.getCookieValue("gender");
+        this.userAge = this.cookieService.getCookieValue("userAge");
+        this.deviceType = this.cookieService.getCookieValue("deviceType");
```
Comment (eldrickb) [inline ui/src/VoxettaUserForm.js:80]: why is this necessary? I thought originally we don't show the form at all if a session cookie exists?or will we just always display it for now?

Comment (ashleygranitto) [reply ui/src/VoxettaUserForm.js:80]: This is for auto-populating when the user accesses the form via the user icon button on the record page in case they want to see their information/make any changes--right now there is no functionality for displaying the form from the beginning/skipping if it is already filled--will complete in a later pr

File: ui/src/VoxettaUserForm.js [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+            </section>
+        `;
+    }
+
+    /**
+     * Checks to see if the user submitted valid information, and if so,
+     * saves the user information in a cookie. 
+     */
+    processForm() {
+        this.userId = this.shadowRoot.getElementById('user-id').value;
```
Comment (mmallett) [inline ui/src/VoxettaUserForm.js:185]: mwc-textfield has built in validation https://github.com/material-components/material-components-web-components/tree/master/packages/textfield#validation

So you can use the default validators or attach custom validation logic to each input field, and also use the `validationMessage` to attach the error directly to that input. Each input will then report its `validity.valid` which you can roll up into one boolean with JavaScript

https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation

File: ui/src/VoxettaUserForm.js [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+            mwc-button.cancel {
+                --mdc-theme-primary: white;
+                --mdc-theme-on-primary: #1a73e8;
+            }
+        `;
+    }
+
+    constructor() {
+        super();
+        this.cookieService = new CookieService();
```
Comment (mmallett) [inline ui/src/VoxettaUserForm.js:106]: Pass these information in from property bindings, not from the cookie. The `vox-app` knows about cookies, but `vox-user-form` doesn't necessarily need to, it can receive this information from the parent.

https://lit-element.polymer-project.org/guide/properties

And conversely, `vox-user-form` can communicate updates back up to parent by dispatching custom events.

https://lit-element.polymer-project.org/guide/events#fire-event

The goal is to minimize responsibilities of each component. Here, we want to isolate cookie dependency to only `vox-app` if possible, and make `vox-user-form` only care about form validation and input management.

File: ui/src/VoxettaUserIcon.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            </button>
+        `;
+    }
+
+    /**
+     * Emits an event that causes the record page to close and the user
+     * form to appear. 
+     */
+    handleUserIconClick() {
+        let event = new CustomEvent('enter-form', {
```
Comment (mmallett) [inline ui/src/VoxettaUserIcon.js:41]: const

File: ui/src/VoxettaUserIcon.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import {Icon} from '@material/mwc-icon';
+import {IconButton} from '@material/mwc-icon-button';
+
+/**
+ * Button responsible for enabling the user to record and upload audio files. 
+ */
+export class VoxettaUserIcon extends LitElement {
+
+    render() {
+        // TO DO (ashley) : stylize and add an actual icon
```
Comment (mmallett) [inline ui/src/VoxettaUserIcon.js:27]: nit: // TODO(ashley):

File: ui/src/VoxettaUserForm.js [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+            return true; 
+        }
+    }
+
+    /**
+     * Emits an event that causes the form page to close and the record
+     * page to appear. 
+     */
+    handleExitForm() {
+        let event = new CustomEvent('exit-form', {
```
Comment (mmallett) [inline ui/src/VoxettaUserForm.js:246]: const

File: ui/src/CookieService.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+     */
+    getCookieValue(key) {
+        const cookie = document.cookie; 
+        if (cookie.includes(key+"=")) {
+            return cookie
+                .split('; ')
+                .find(entry => entry.startsWith(key))
+                .split('=')[1];
+        } else {
+            return '';
```
[the remark this answers is not in the document]
Comment (ashleygranitto) [reply ui/src/CookieService.js:47]: this was intentional because of the auto-population feature--if a cookie does not exist, I want the input to be initially blank & they are set through this function

File: ui/src/CookieService.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+ 
+/**
+ * Service responsible for creating and fetching cookies.   
+ */
+export class CookieService {
```
[the remark this answers is not in the document]
Comment (ashleygranitto) [reply ui/src/CookieService.js:19]: I have not done any testing yet--I was hoping to do them in a later pull request after I am able to get Elvis's Jasmine flow from Master

File: ui/src/CookieService.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+     * Create a cookie for each component of user information.  
+     * @param {String} key - The name of a cookie key whose value is sought.
+     * @returns {String} The value of the given cookie key, or an empty string
+     *  if a cookie has not been set for such a key.  
+     */
+    getCookieValue(key) {
+        const cookie = document.cookie; 
+        if (cookie.includes(key+"=")) {
```
[the remark this answers is not in the document]
Comment (ashleygranitto) [reply ui/src/CookieService.js:41]: fixed

File: ui/src/VoxettaUserIcon.js (hunk 6 above)
Comment (ashleygranitto) [reply ui/src/VoxettaUserIcon.js:27]: forgot to remove after I added the icon--now removed--but will remember this style in the future

File: ui/src/VoxettaUserForm.js (hunk 7 above)
Comment (ashleygranitto) [reply ui/src/VoxettaUserForm.js:246]: adjusted

File: ui/src/VoxettaUserIcon.js (hunk 5 above)
Comment (ashleygranitto) [reply ui/src/VoxettaUserIcon.js:41]: adjusted

File: ui/src/VoxettaApp.js [hunk 11]
```diff
@@ -13,18 +13,53 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+
+    static get properties() {
+        return {
+            state: {type: String}
+        };
+    }
+
+    constructor() {
+        super();
+        this.state = 'RECORD-PAGE';
```
[the remark this answers is not in the document]
Comment (ashleygranitto) [reply ui/src/VoxettaApp.js:32]: adjusted--wasn't entirely sure if I should have the States constant in each applicable file or just the main app and adjust the event logic accordingly. As of now I have the States constant in each applicable file but please let me know if I should change this

File: ui/src/VoxettaUserForm.js (hunk 4 above)
Comment (ashleygranitto) [reply ui/src/VoxettaUserForm.js:106]: adjusted

File: ui/src/VoxettaUserForm.js (hunk 3 above)
Comment (ashleygranitto) [reply ui/src/VoxettaUserForm.js:192]: fixed

File: ui/src/VoxettaApp.js [hunk 12]
```diff
@@ -29,13 +38,17 @@ export class VoxettaApp extends LitElement {
 
     constructor() {
         super();
-        this.state = 'RECORD-PAGE'; 
+        this.cookieService = new CookieService();
+        this.state = States.RECORD_PAGE; 
+        this.userId = this.cookieService.getCookieValue("userId");
```
Comment (mmallett) [inline ui/src/VoxettaApp.js:43]: suggestion, cookie service should provide getters for each supported field.

such as:

```javascript
this.cookieService.getUserId();
this.cookieService.getGender();
this.cookieService.getUserAge();
```

So the consumer of cookie service doesn't need to know the exact names of the fields

File: ui/src/VoxettaUserForm.js [hunk 13]
```diff
@@ -187,69 +201,53 @@ export class VoxettaUserForm extends LitElement {
...
-            this.ensureInputCompleteness('user-id', 'Please enter a valid user id.');
-        const genderComplete =
-            this.ensureInputCompleteness('gender-list', 'Please select a gender.');
-        const ageComplete =
-            this.ensureInputCompleteness('user-age', 
-                'Please enter a valid age between 0 and 120.');
-        const deviceComplete =
-            this.ensureInputCompleteness('device-type', 'Please enter a device name.');
+    formIsValid() {
+        const userIdValidity = this.shadowRoot.getElementById('user-id').checkValidity();
```
Comment (mmallett) [inline ui/src/VoxettaUserForm.js:219]: can maybe squish this a bit

```javascript
const form = this.shadowRoot.getElementById('my-cool-form');
const valid = form.elements.every((element) => {
  return element.checkValidity();
});
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every
https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/elements

File: ui/src/VoxettaUserForm.js [hunk 14]
```diff
@@ -187,69 +201,53 @@ export class VoxettaUserForm extends LitElement {
...
+     * Emits an event that causes the user infomration to update. 
+     * @param {Object} userInfo - The information entered on the user form.
      */
-    handleExitForm() {
-        let event = new CustomEvent('exit-form', {
+    handleFormSubmission(userInfo) {
+        const event = new CustomEvent('update-user-info', {
             detail: {
-                state: 'RECORD-PAGE'
+                userInfo: userInfo
```
Comment (mmallett) [inline ui/src/VoxettaUserForm.js:246]: you can write this simply

```javascript
detail: { userInfo }
```

If key is the same as the variable set as value, this applies. See shorthand object notation
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer

File: ui/src/VoxettaUserForm.js [hunk 15]
```diff
@@ -103,11 +111,7 @@ export class VoxettaUserForm extends LitElement {
 
     constructor() {
         super();
-        this.cookieService = new CookieService();
-        this.userId = this.cookieService.getCookieValue("userId");
-        this.gender = this.cookieService.getCookieValue("gender");
-        this.userAge = this.cookieService.getCookieValue("userAge");
-        this.deviceType = this.cookieService.getCookieValue("deviceType"); 
+        this.addEventListener('input', this.formIsValid);
```
Comment (mmallett) [inline ui/src/VoxettaUserForm.js:114]: I think `mwc-select` fires `@selected` not `@input`, so this wouldn't catch changes to gender.

File: ui/src/VoxettaUserForm.js [hunk 16]
```diff
@@ -187,69 +201,53 @@ export class VoxettaUserForm extends LitElement {
...
+        const userIdValidity = this.shadowRoot.getElementById('user-id').checkValidity();
+        const genderValidity = this.shadowRoot.getElementById('gender-list').checkValidity();
+        const userAgeValidity = this.shadowRoot.getElementById('user-age').checkValidity();
+        const deviceTypeValidity = this.shadowRoot.getElementById('device-type').checkValidity();
 
-        return idComplete && genderComplete && ageComplete && deviceComplete;  
+        const formValidity = userIdValidity && genderValidity && userAgeValidity && deviceTypeValidity;
+        
+        const saveButton = this.shadowRoot.getElementById('save-button');
+        formValidity ? saveButton.disabled  = false : saveButton.disabled = true;
```
Comment (mmallett) [inline ui/src/VoxettaUserForm.js:227]:
```javascript
saveButton.disable = formValidity;
```

File: ui/src/VoxettaUserForm.js [hunk 17]
```diff
@@ -28,6 +27,15 @@ import {Icon} from '@material/mwc-icon'
  */
 export class VoxettaUserForm extends LitElement {
 
+    static get properties() {
+        return {
+            userId: {type: String},
+            gender: {type: String},
+            userAge: {type: String},
```
Comment (mmallett) [inline ui/src/VoxettaUserForm.js:34]: age is a `Number`

File: ui/src/CookieService.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (eldrickb) [reply ui/src/CookieService.js]: fyi  `npm run lint` will fix small things like this for you

File: ui/src/VoxettaUserForm.js (hunk 17 above)
Comment (ashleygranitto) [reply ui/src/VoxettaUserForm.js:34]: fixed

File: ui/src/VoxettaUserForm.js (hunk 16 above)
Comment (ashleygranitto) [reply ui/src/VoxettaUserForm.js:227]: adjusted

File: ui/src/VoxettaUserForm.js (hunk 14 above)
Comment (ashleygranitto) [reply ui/src/VoxettaUserForm.js:246]: adjusted

File: ui/src/VoxettaUserForm.js (hunk 13 above)
Comment (ashleygranitto) [reply ui/src/VoxettaUserForm.js:219]: I unfortunately don't think the .elements function is applicable because I am not using the components that this function captures, I also tried using .childNodes but this led to an issue with the select div (necessary style workaround--currently there is know bug with `<mwc-select>` width)

File: ui/src/VoxettaApp.js (hunk 12 above)
Comment (ashleygranitto) [reply ui/src/VoxettaApp.js:43]: adjusted

File: ui/src/VoxettaUserForm.js (hunk 15 above)
Comment (ashleygranitto) [reply ui/src/VoxettaUserForm.js:114]: Oh good catch! I just tried using 'onselect' and 'onchange' but it doesn't seem to recognize either--I'm getting errors both creating an independent event & putting it within the HTML--I ultimately used an additional independent 'onclick' event (I figured the user will have to click to make a selection) but this is less ideal because it makes some unnecessary checks. I will revisit in the future to see if I can get an 'onchange' variation working
